add strobelight cli function profiler

laithsakka opened the pull request:
This is a meta only tool, this allow users to profile any python function by annotating it with **strobelight** using 
the strobelight profiler. 
ex
```
    def fn(x, y, z):
        return x * y + z

    # use decorator with default profiler.
    @strobelight()
    @torch.compile()
    def work():
        for i in range(100):
            for j in range(5):
                fn(torch.rand(j, j), torch.rand(j, j), torch.rand(j, j))

    work()
```

test
```
 python torch/utils/strobelight/examples/cli_function_profiler_example.py 
strobelight_cli_function_profiler, line 274, 2024-05-20 11:05:41,513, INFO: strobelight run id is: -6222660165281106
strobelight_cli_function_profiler, line 276, 2024-05-20 11:06:08,318, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 257, 2024-05-20 11:06:11,867, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:16,164, INFO: Total samples: 2470
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:16,164, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/oiqmyltg
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:16,164, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/b10x92x0
strobelight_cli_function_profiler, line 274, 2024-05-20 11:06:18,476, INFO: strobelight run id is: -4112659701221677
strobelight_cli_function_profiler, line 276, 2024-05-20 11:06:45,096, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 257, 2024-05-20 11:06:52,366, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:56,222, INFO: Total samples: 1260
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:56,222, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/0yyx6el5
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:56,223, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/8m2by4ea
(base) [lsakka@devvm4561.ash0 /data/users/lsakka/pytorch/pytorch (strobelight2)]$ python torch/profiler/strobelight_cli_function_profiler_example.py 
strobelight_cli_function_profiler, line 274, 2024-05-20 11:07:26,701, INFO: strobelight run id is: -2373009368202256
strobelight_cli_function_profiler, line 276, 2024-05-20 11:07:53,477, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 257, 2024-05-20 11:07:56,827, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:01,138, INFO: Total samples: 2372
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:01,138, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/dk797xg9
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:01,138, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/4w6c8vnm
strobelight_cli_function_profiler, line 274, 2024-05-20 11:08:03,235, INFO: strobelight run id is: -1919086123693716
strobelight_cli_function_profiler, line 276, 2024-05-20 11:08:29,848, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 257, 2024-05-20 11:08:37,233, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:41,138, INFO: Total samples: 1272
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:41,138, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/43r58aew
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:41,138, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/9g52onmw
(base) [lsakka@devvm4561.ash0 /data/users/lsakka/pytorch/pytorch (strobelight2)]$ 
```


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126693

laithsakka committed:
add strobelight cli function profiler

[ghstack-poisoned]

laithsakka committed:
Update on "add strobelight cli function profiler"

This is a meta only tool, this allow users to profile any python function by annotating it with **strobelight** using 
the strobelight profiler. 
ex
```
    def fn(x, y, z):
        return x * y + z

    # use decorator with default profiler.
    strobelight()
    torch.compile()
    def work():
        for i in range(100):
            for j in range(5):
                fn(torch.rand(j, j), torch.rand(j, j), torch.rand(j, j))

    work()
```

test
```
(base) [lsakkadevvm4561.ash0 /data/users/lsakka/pytorch/pytorch (strobelight2)]$ python torch/profiler/strobelight_cli_function_profiler_example.py 
strobelight_cli_function_profiler, line 285, 2024-05-20 09:45:49,436, INFO: strobelight run id is: -4310015838269154
strobelight_cli_function_profiler, line 287, 2024-05-20 09:46:06,200, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:46:08,300, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: Total samples: 10
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/ad9ec2h1
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/i9cftjh0
strobelight_cli_function_profiler, line 285, 2024-05-20 09:46:14,445, INFO: strobelight run id is: 7449419640850128
strobelight_cli_function_profiler, line 287, 2024-05-20 09:46:37,852, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:46:44,804, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: Total samples: 1242
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/5fw5i3f5
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/f6y5pahz
strobelight_cli_function_profiler, line 285, 2024-05-20 09:46:50,160, INFO: strobelight run id is: -6100585386049331
strobelight_cli_function_profiler, line 287, 2024-05-20 09:47:13,285, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:47:25,284, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: Total samples: 2
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/e0stqsbr
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/tw7r8jte
```


[ghstack-poisoned]

laithsakka committed:
Update on "add strobelight cli function profiler"

This is a meta only tool, this allow users to profile any python function by annotating it with **strobelight** using 
the strobelight profiler. 
ex
```
    def fn(x, y, z):
        return x * y + z

    # use decorator with default profiler.
    strobelight()
    torch.compile()
    def work():
        for i in range(100):
            for j in range(5):
                fn(torch.rand(j, j), torch.rand(j, j), torch.rand(j, j))

    work()
```

test
```
(base) [lsakkadevvm4561.ash0 /data/users/lsakka/pytorch/pytorch (strobelight2)]$ python torch/profiler/strobelight_cli_function_profiler_example.py 
strobelight_cli_function_profiler, line 285, 2024-05-20 09:45:49,436, INFO: strobelight run id is: -4310015838269154
strobelight_cli_function_profiler, line 287, 2024-05-20 09:46:06,200, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:46:08,300, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: Total samples: 10
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/ad9ec2h1
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/i9cftjh0
strobelight_cli_function_profiler, line 285, 2024-05-20 09:46:14,445, INFO: strobelight run id is: 7449419640850128
strobelight_cli_function_profiler, line 287, 2024-05-20 09:46:37,852, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:46:44,804, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: Total samples: 1242
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/5fw5i3f5
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/f6y5pahz
strobelight_cli_function_profiler, line 285, 2024-05-20 09:46:50,160, INFO: strobelight run id is: -6100585386049331
strobelight_cli_function_profiler, line 287, 2024-05-20 09:47:13,285, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:47:25,284, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: Total samples: 2
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/e0stqsbr
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/tw7r8jte
```


[ghstack-poisoned]

laithsakka committed:
Update on "add strobelight cli function profiler"

This is a meta only tool, this allow users to profile any python function by annotating it with **strobelight** using 
the strobelight profiler. 
ex
```
    def fn(x, y, z):
        return x * y + z

    # use decorator with default profiler.
    strobelight()
    torch.compile()
    def work():
        for i in range(100):
            for j in range(5):
                fn(torch.rand(j, j), torch.rand(j, j), torch.rand(j, j))

    work()
```

test
```
(base) [lsakkadevvm4561.ash0 /data/users/lsakka/pytorch/pytorch (strobelight2)]$ python torch/profiler/strobelight_cli_function_profiler_example.py 
strobelight_cli_function_profiler, line 285, 2024-05-20 09:45:49,436, INFO: strobelight run id is: -4310015838269154
strobelight_cli_function_profiler, line 287, 2024-05-20 09:46:06,200, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:46:08,300, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: Total samples: 10
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/ad9ec2h1
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/i9cftjh0
strobelight_cli_function_profiler, line 285, 2024-05-20 09:46:14,445, INFO: strobelight run id is: 7449419640850128
strobelight_cli_function_profiler, line 287, 2024-05-20 09:46:37,852, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:46:44,804, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: Total samples: 1242
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/5fw5i3f5
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/f6y5pahz
strobelight_cli_function_profiler, line 285, 2024-05-20 09:46:50,160, INFO: strobelight run id is: -6100585386049331
strobelight_cli_function_profiler, line 287, 2024-05-20 09:47:13,285, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:47:25,284, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: Total samples: 2
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/e0stqsbr
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/tw7r8jte
```


[ghstack-poisoned]

laithsakka committed:
Update on "add strobelight cli function profiler"

This is a meta only tool, this allow users to profile any python function by annotating it with **strobelight** using 
the strobelight profiler. 
ex
```
    def fn(x, y, z):
        return x * y + z

    # use decorator with default profiler.
    strobelight()
    torch.compile()
    def work():
        for i in range(100):
            for j in range(5):
                fn(torch.rand(j, j), torch.rand(j, j), torch.rand(j, j))

    work()
```

test
```
(base) [lsakkadevvm4561.ash0 /data/users/lsakka/pytorch/pytorch (strobelight2)]$ python torch/profiler/strobelight_cli_function_profiler_example.py 
strobelight_cli_function_profiler, line 285, 2024-05-20 09:45:49,436, INFO: strobelight run id is: -4310015838269154
strobelight_cli_function_profiler, line 287, 2024-05-20 09:46:06,200, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:46:08,300, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: Total samples: 10
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/ad9ec2h1
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:12,233, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/i9cftjh0
strobelight_cli_function_profiler, line 285, 2024-05-20 09:46:14,445, INFO: strobelight run id is: 7449419640850128
strobelight_cli_function_profiler, line 287, 2024-05-20 09:46:37,852, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:46:44,804, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: Total samples: 1242
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/5fw5i3f5
strobelight_cli_function_profiler, line 230, 2024-05-20 09:46:48,121, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/f6y5pahz
strobelight_cli_function_profiler, line 285, 2024-05-20 09:46:50,160, INFO: strobelight run id is: -6100585386049331
strobelight_cli_function_profiler, line 287, 2024-05-20 09:47:13,285, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 248, 2024-05-20 09:47:25,284, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: Total samples: 2
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/e0stqsbr
strobelight_cli_function_profiler, line 230, 2024-05-20 09:47:29,250, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/tw7r8jte
```


[ghstack-poisoned]

laithsakka committed:
Update

[ghstack-poisoned]

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+
+
+class StrobelightCLIProfilerError(Exception):
+    "Raised when an error happens during strobelight profiling"
```
I believe that any string will be treated as a doc comment - `"""` is just convention (and allows multi-line docs containing `"`).  Also the string acts as a statement so `pass` is unnecessary (but would probably make it easier to read)

laithsakka replied:
i will use 
"""
"""

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+    return int(link[link.find("[") + 1 : -1])
+
+
+def command_to_string(command):
```
```
def command_to_string(command: Sequence[str]) -> str:
```

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+
+
+def command_to_string(command):
+    s = ""
+    for item in command:
+        s += " "
+        s += item
+    return s
```
nit: this whole thing is just this and can be inlined:
```
return " " + " ".join(command)
```
(or if you didn't mean to always have a leading space then:
```
return " ".join(command)
```
)

laithsakka commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+from threading import Lock
+from typing import Optional
+
+logger = logging.getLogger("strobelight_cli_function_profiler")
```
I am not sure we do have strobelight_function_profiler under torch/fb .

laithsakka commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+# A function decorator that wraps profile, if no profiler is provided one with
+# default args is created. A function can be annotated as:
+# @strobelight()
+# @strobelight(profiler = StrobelightFunctionProfiler(stop_at_error=True,..))
```
what do you mean?
never you can either do 
strobelight(profiler args)
or strobelight(profiler object)
or strobelight()

laithsakka commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+
+            self._get_results()
+        except Exception as error:
+            logger.warning("error during stop_strobelight %s", error)
```
can you comment what does that do?

laithsakka replied:
https://stackoverflow.com/questions/192261/how-do-i-log-an-exception-at-warning-or-info-level-with-traceback-using-the-pyt

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka commented:
Address comments

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+            "cycles",
+            "--async",
+            "--sample-interval",
+            str(int(self.sample_each)),
```
nit:
```
f"{self.sample_each}",
```

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+            "--sample-interval",
+            str(int(self.sample_each)),
+            "--duration-ms",
+            str(int(self.max_profile_duration_sec * 1000)),
```
nit:
```
f"{self.max_profile_duration_sec * 1000}",
```

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+        ]
+
+        if self.sample_tags:
+            command.append("--sample-tags")
+            add_comma = False
+            for item in self.sample_tags:
+                if add_comma:
+                    command.append(",")
+                command.append(item)
+                add_comma = True
```
This can't be right - you're adding "," as a separate command argument -- the command will be something like:
```
"strobeclient" "run" ... "--sample-tags" "foo" "," "bar" "," "baz"
```
You probably want something like:
```
command.append("--sample-tags")
command.append(",".join(self.sample_tags))
```
So the sample tags are added as a single argument.

laithsakka replied:
yeh that i did not test since it it will be tested as part of the following diff, i will address it and maybe add example with sample tags to test as well part of this diff.

laithsakka replied:
let me check i though it is "strobeclient" "run" ... "--sample-tags" "foo" "," "bar" "," "baz"

laithsakka replied:
ah you are right no spaces should be there, i addressed it and added sample_tags in one example

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+                "failed to start strobelight profiling, error in run_async"
+            )
+
+        for line in result.stderr.split(b"\n"):
```
nit:
```
for line in result.stderr.splitlines():
```

but less nit:
Replace this whole `for` loop with a single `re.search` for the pattern you're looking for - which should include the match group - something like this:
```
if match := re.search(rb"INFO Run Id:.*(-?\d+)$", line):
  self.current_run_id = int(match.group(1))
```

laithsakka replied:
mhmm let me look i think the reason i added the loop is that logs get pritned all in one line (stream of bytes) but I can decode result first to avoid the loop.

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+                "wait_for_running called more than 20 times"
+            )
+
+        command = ["strobeclient", "getRunStatus", "--run-id", str(self.current_run_id)]
```
nit:
```
"f{self.current_run_id}"
```

laithsakka committed:
Update

[ghstack-poisoned]

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+                "failed to start strobelight profiling, error in wait_for_running"
+            )
+
+        for line in result.stderr.split(b"\n"):
```
Same deal as above - no need to split and search by lines - just search in the whole unsplit stderr.

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+                "failed to stop strobelight profiling, return code is not 0"
+            )
+
+        for line in result.stderr.split(b"\n"):
```
again - don't split

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+                "failed to extract profiling results, return code is not 0"
+            )
+
+        for line in result.stderr.split(b"\n"):
```
here too - no split needed

aorenste commented on torch/profiler/strobelight_cli_function_profiler.py:
```diff
@@ -0,0 +1,311 @@
+import functools
```
Every function should have types because mypy doesn't typecheck functions with missing types.  
If you want to enforce this for the whole file you can add to the first line:
```
# mypy: disallow-untyped-defs
```

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka committed:
Update on "add strobelight cli function profiler"

This is a meta only tool, this allow users to profile any python function by annotating it with **strobelight** using 
the strobelight profiler. 
ex
```
    def fn(x, y, z):
        return x * y + z

    # use decorator with default profiler.
    strobelight()
    torch.compile()
    def work():
        for i in range(100):
            for j in range(5):
                fn(torch.rand(j, j), torch.rand(j, j), torch.rand(j, j))

    work()
```

test
```
$ python torch/profiler/strobelight_cli_function_profiler_example.py 
strobelight_cli_function_profiler, line 274, 2024-05-20 11:05:41,513, INFO: strobelight run id is: -6222660165281106
strobelight_cli_function_profiler, line 276, 2024-05-20 11:06:08,318, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 257, 2024-05-20 11:06:11,867, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:16,164, INFO: Total samples: 2470
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:16,164, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/oiqmyltg
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:16,164, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/b10x92x0
strobelight_cli_function_profiler, line 274, 2024-05-20 11:06:18,476, INFO: strobelight run id is: -4112659701221677
strobelight_cli_function_profiler, line 276, 2024-05-20 11:06:45,096, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 257, 2024-05-20 11:06:52,366, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:56,222, INFO: Total samples: 1260
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:56,222, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/0yyx6el5
strobelight_cli_function_profiler, line 237, 2024-05-20 11:06:56,223, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/8m2by4ea
(base) [lsakkadevvm4561.ash0 /data/users/lsakka/pytorch/pytorch (strobelight2)]$ python torch/profiler/strobelight_cli_function_profiler_example.py 
strobelight_cli_function_profiler, line 274, 2024-05-20 11:07:26,701, INFO: strobelight run id is: -2373009368202256
strobelight_cli_function_profiler, line 276, 2024-05-20 11:07:53,477, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 257, 2024-05-20 11:07:56,827, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:01,138, INFO: Total samples: 2372
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:01,138, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/dk797xg9
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:01,138, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/4w6c8vnm
strobelight_cli_function_profiler, line 274, 2024-05-20 11:08:03,235, INFO: strobelight run id is: -1919086123693716
strobelight_cli_function_profiler, line 276, 2024-05-20 11:08:29,848, INFO: strobelight profiling running
strobelight_cli_function_profiler, line 257, 2024-05-20 11:08:37,233, INFO: strobelight profiling stopped
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:41,138, INFO: Total samples: 1272
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:41,138, INFO: GraphProfiler (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/43r58aew
strobelight_cli_function_profiler, line 237, 2024-05-20 11:08:41,138, INFO: Icicle view (python stack): https://fburl.com/scuba/pyperf_experimental/on_demand/9g52onmw
(base) [lsakkadevvm4561.ash0 /data/users/lsakka/pytorch/pytorch (strobelight2)]$ 
```


[ghstack-poisoned]

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka committed:
Update

[ghstack-poisoned]

aorenste commented on tools/strobelight/cli_function_profiler.py:
```diff
@@ -0,0 +1,307 @@
+from threading import Lock
+from typing import Any, List, Optional, Sequence
+
+# mypy: disallow-untyped-defs
```
I think the convention is usually these go at the very top of the file since they apply to the whole file.

aorenste commented on tools/strobelight/cli_function_profiler.py:
```diff
@@ -0,0 +1,307 @@
+
+def _pid_namespace_link(pid: Optional[int] = None) -> str:
+    PID_NAMESPACE_PATH = "/proc/{}/ns/pid"
+    """Returns the link to the process's namespace, example: pid:[4026531836]"""
```
I think the docstring needs to be the first line of the function: https://peps.python.org/pep-0257/#what-is-a-docstring

laithsakka replied:
I copy pasted this function from an internal meta file i can move it

aorenste commented on tools/strobelight/cli_function_profiler.py:
```diff
@@ -0,0 +1,307 @@
+
+class StrobelightCLIFunctionProfiler:
+    """
+    Note: this is a meta only tool.
```
capital 'M'eta.

laithsakka replied:
ah i missed this will address it innext one

aorenste commented on tools/strobelight/cli_function_profiler.py:
```diff
@@ -0,0 +1,307 @@
+                f"failed to start strobelight profiling, error in run_async:{output}"
+            )
+
+        match = re.search("INFO Run Id:.*", output)
+        if match:
+            inner_match = re.search(r"-?\d+$", match.group(0))
+            if inner_match:
+                self.current_run_id = int(inner_match.group(0))
+                return
```
Nit: This could be written as:
```
if match := re.search("INFO Run Id:.*(-?\d+)$", output, re.M):
  self.current_run_id = int(match.group(1))
  return
```

laithsakka replied:
i will address this and use if match = in all other places

aorenste commented on tools/strobelight/cli_function_profiler.py:
```diff
@@ -0,0 +1,307 @@
+            else:
+                raise StrobelightCLIProfilerError(f"unexpected {current_status} phase")
+
+        raise StrobelightCLIProfilerError("unreachable")
```
Technically not unreachable - just an unexpected output from strobelight.

laithsakka replied:
ok will change that

aorenste commented on tools/strobelight/cli_function_profiler.py:
```diff
@@ -0,0 +1,307 @@
+                    f"failed to stop strobelight profiling, got {current_status} result"
+                )
+
+        raise StrobelightCLIProfilerError("unreachable")
```
Technically not unreachable - just an unexpected output from strobelight.

aorenste commented on tools/strobelight/cli_function_profiler.py:
```diff
@@ -0,0 +1,307 @@
+    def profile(self, work_function: Any, *args: Any, **kwargs: Any) -> Any:
+        self.current_run_id = None
+
+        if StrobelightCLIFunctionProfiler._lock.locked():
```
This is actually a race condition - if the lock is locked between here and line 269 then you'll wait for the lock rather than complain/run the work function. But it's really unlikely I think... If we wanted to fix it the proper way to deal with it is to use _lock.acquire(False) and check the return value.

laithsakka replied:
I see what you are saying.

laithsakka replied:
```
    if locked := StrobelightCLIFunctionProfiler._lock.acquire(False):
            if not locked:
                if self.stop_at_error:
                    raise StrobelightCLIProfilerError("simultaneous runs not supported")

                return work_function(*args, **kwargs)
            
            started = self._start_strobelight()
            if not started:
                if self.stop_at_error:
                    raise StrobelightCLIProfilerError(
                        "failed to start strobelight profiling"
                    )
                return work_function(*args, **kwargs)

            try:
                logger.debug("collection started")
                result = work_function(*args, **kwargs)
                self._stop_strobelight_no_throw(collect_results=True)
                return result
            except Exception as error:
                logger.warning("work function throw exception", exc_info=True)
                self._stop_strobelight_no_throw(collect_results=False)
                raise error
```

laithsakka replied:
addressed

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka commented:
@pytorchbot  merge

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka commented:
moving to torch/utils. seems like there is issues of torch depending on /tools/

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka committed:
Update

[ghstack-poisoned]

laithsakka commented:
make _strobelight private, since its only used by meta engineers. does not make sense to have public documentation for it

laithsakka commented:
@pytorchbot merge
